mantaray/received.py

```diff
@@ -312,12 +332,27 @@ def _handle_cap(server_view: views.ServerView, args: list[str]) -> None:
     if subcommand == "ACK":
         acknowledged = set(args[-1].split())
         if "sasl" in acknowledged:
+            server_view.core.cap_list.add("sasl")
             server_view.core.send("AUTHENTICATE PLAIN")
+        if "away-notify" in acknowledged:
+            server_view.core.cap_list.add("away-notify")
     elif subcommand == "NAK":
         rejected = set(args[-1].split())
         if "sasl" in rejected:
             # TODO: this good?
             raise ValueError("The server does not support SASL.")
+        if "away-notify" in rejected:
+            raise ValueError("The server does not support away-notify.")
+
+    server_view.core.cap_req.pop()  # To evaluate how many more ACK/NAKs will be received from server
+
+    # Need to wait for login success/fail from server if sasl is enabled before sending "CAP END"
+    if len(server_view.core.cap_req) == 0 and "sasl" not in server_view.core.cap_list:
+        _send_cap_end(server_view)
+
+
+def _send_cap_end(server_view: views.ServerView) -> None:
+    server_view.core.send("CAP END")
```

Comment:
I don't like short functions like this. Feel free to write them into your own projects, but in my projects, if there's a one-line send, I want to just have that one-line send, not hidden inside a function.

Also applies to `send_who()` method in `backend.py`.